tsdb/mapper.go

```diff
@@ -478,71 +529,80 @@ func (tsc *tagSetCursor) key() string {
-		if min == -1 {
+		// If we're out of points, we're done.
+		if tsc.pqueue.Len() == 0 {
 			return -1, nil
 		}
 
-		// set the current timestamp and seriesID
-		timestamp := tsc.keyBuffer[min]
+		// Grab the next point with the lowest timestamp.
+		p := heap.Pop(tsc.pqueue).(*rawPoint)
 
-		var value interface{}
-		if len(selectFields) > 1 {
-			if fieldsWithNames, err := tsc.decoder.DecodeFieldsWithNames(tsc.valueBuffer[min]); err == nil {
-				value = fieldsWithNames
+		// We're done if the point is outside the query's time range [tmin:tmax).
+		if p.timestamp != tmin && (tmin > p.timestamp || p.timestamp >= tmax) {
+			return -1, nil
+		}
 
-				// if there's a where clause, make sure we don't need to filter this value
-				if tsc.cursors[min].filter != nil && !matchesWhere(tsc.cursors[min].filter, fieldsWithNames) {
-					value = nil
-				}
-			}
-		} else {
-			// With only 1 field SELECTed, decoding all fields may be avoidable, which is faster.
-			var err error
-			value, err = tsc.decoder.DecodeByName(selectFields[0], tsc.valueBuffer[min])
-			if err != nil {
-				value = nil
-			} else {
-				// If there's a WHERE clase, see if we need to filter
-				if tsc.cursors[min].filter != nil {
-					// See if the WHERE is only on this field or on one or more other fields.
-					// If the latter, we'll have to decode everything
-					if len(whereFields) == 1 && whereFields[0] == selectFields[0] {
-						if !matchesWhere(tsc.cursors[min].filter, map[string]interface{}{selectFields[0]: value}) {
-							value = nil
-						}
-					} else { // Decode everything
-						fieldsWithNames, err := tsc.decoder.DecodeFieldsWithNames(tsc.valueBuffer[min])
-						if err != nil || !matchesWhere(tsc.cursors[min].filter, fieldsWithNames) {
-							value = nil
-						}
-					}
-				}
+		// Advance the cursor
+		nextKey, nextVal := p.cursor.Next()
+		if nextKey != -1 {
+			nextPoint := &rawPoint{
+				timestamp: nextKey,
+				value:     nextVal,
+				cursor:    p.cursor,
 			}
+			heap.Push(tsc.pqueue, nextPoint)
 		}
 
-		// Advance the cursor
-		nextKey, nextVal := tsc.cursors[min].Next()
-		tsc.keyBuffer[min] = nextKey
-		tsc.valueBuffer[min] = nextVal
+		// Decode the raw point.
+		value := tsc.decodeRawPoint(p, selectFields, whereFields)
 
 		// Value didn't match, look for the next one.
 		if value == nil {
 			continue
 		}
 
-		return timestamp, value
+		return p.timestamp, value
+	}
+}
+
+// decodeRawPoint decodes raw point data into field names & values and does WHERE filtering.
+func (tsc *tagSetCursor) decodeRawPoint(p *rawPoint, selectFields, whereFields []string) interface{} {
+	var value interface{}
+	if len(selectFields) > 1 {
+		if fieldsWithNames, err := tsc.decoder.DecodeFieldsWithNames(p.value); err == nil {
+			value = fieldsWithNames
+
+			// if there's a where clause, make sure we don't need to filter this value
+			if p.cursor.filter != nil && !matchesWhere(p.cursor.filter, fieldsWithNames) {
+				value = nil
+			}
+		}
+	} else {
```

Comment:
Can you `return` from this `if` block so that the `else` part can be removed and it'll flatten the indentation a bit?